src/util/settings.cpp

```diff
@@ -0,0 +1,161 @@
+    Source& SetConfigFile(bool top_level) { config_file = true; config_file_top_level = top_level; return *this; }
+};
+
+template <typename Fn>
+static void MergeSettings(const Settings& settings, const std::string& section, const std::string& name, Fn&& fn)
+{
+    if (auto* value = FindKey(settings.forced_settings, name)) {
+        fn(Source(SettingsSpan(*value)).SetForced());
+    }
+    if (auto* values = FindKey(settings.command_line_options, name)) {
+        fn(Source(SettingsSpan(*values)));
+    }
+    if (!section.empty()) {
+        if (auto* map = FindKey(settings.ro_config, section)) {
+            if (auto* values = FindKey(*map, name)) {
+                fn(Source(SettingsSpan(*values)).SetConfigFile(/* top_level= */ false));
+            }
+        }
+    }
+    SettingsSpan span;
+    if (auto* map = FindKey(settings.ro_config, "")) {
+        if (auto* values = FindKey(*map, name)) {
+            span = SettingsSpan(*values);
+        }
+    }
+    fn(Source(span).SetConfigFile(/* top_level= */ true));
+}
+} // namespace
+
+SettingsValue GetSetting(const Settings& settings,
+    const std::string& section,
+    const std::string& name,
+    bool ignore_default_section_config,
+    bool get_chain_name)
+{
+    SettingsValue result;
+    MergeSettings(settings, section, name, [&](Source source) {
+        // Weird behavior preserved for backwards compatibility: Apply negated
+        // setting in otherwise ignored sections. A negated value in the top
+        // level section is applied to network specific options, even though
+        // non-negated values there would be ignored.
+        const bool never_ignore_negated_setting = source.span.last_negated();
+
+        // Weird behavior preserved for backwards compatibility: Take first
+        // assigned value instead of last. In general, later settings take
+        // precedence over early settings, but for backwards compatibility in
+        // the config file the precedence is reversed for most settings.
+        const bool reverse_precedence = source.config_file && !get_chain_name;
+
+        // Weird behavior preserved for backwards compatibility: Negated
+        // -regtest and -testnet arguments which you would expect to override
+        // values set in the configuration file are currently accepted but
+        // silently ignored. It would be better to apply these just like other
+        // negated values, or at least warn they are ignored.
+        const bool skip_negated_command_line = get_chain_name;
+
+        // Ignore settings in top-level config section if requested.
+        if (ignore_default_section_config && source.config_file_top_level && !never_ignore_negated_setting) return;
+
+        // Skip negated command line settings.
+        if (skip_negated_command_line && source.span.last_negated()) return;
+
+        // Stick with highest priority value, keeping result if already set.
+        if (!result.isNull()) return;
```

Comment:
This looks redundant. `result` is always null here, I don't see where it can be changed above. All this could be rewritten as a `return <something>` instead of changing via lambda capturer reference...

Reply:
I can add a 

> This looks redundant. `result` is always null here, I don't see where it can be changed above. All this could be rewritten as a `return <something>` instead of changing via lambda capturer reference...

I can maybe change the comment above to clarify, but `return` here is what keeps lower priority values from overwriting higher priority values,

Reply:
Yep, I've missed that fact that lambda is called multiple times, thanks for clarification. So this comment is moot.